lib/puppet/pops/merge_strategy.rb

```diff
@@ -0,0 +1,218 @@
+    #
+    # @param strategy_class [Class<MergeStrategy>] The class of the added strategy
+    #
+    def self.add_strategy(strategy_class)
+      raise ArgumentError, "MergeStrategies.add_strategy 'strategy_class' must be a 'MergeStrategy' class. Got #{strategy_class}" unless MergeStrategy > strategy_class
+      s = strategy_class.new
+      strategies[s.key] = s
+      nil
+    end
+
+    # Finds a merge strategy that corresponds to the given _merge_ argument and delegates the task of merging the elements of _e1_ and _e2_ to it.
+    #
+    # @param e1 [Object] The first element
+    # @param e2 [Object] The second element
+    # @param merge [String|Symbol|Hash<String,Object>] The merge strategy. Can be a string or symbol denoting the key identifier or a hash with options where the key 'strategy' denotes the key
+    # @return [Object] The result of the merge
+    #
+    def self.merge(e1, e2, merge)
+      TypeAsserter.assert_instance_of("MergeStrategies.merge 'merge' parameter", merge_t, merge)
+      if merge.is_a?(Hash)
+        merge_strategy = merge['strategy']
+        raise ArgumentError, "MergeStrategies.merge 'merge' parameter must contain a 'strategy' of type String" if merge_strategy.nil?
+        merge_options  = merge
+      else
+        merge_strategy = merge
+        merge_options  = {}
+      end
+      strategy(merge_strategy.to_sym).merge(e1, e2, merge_options)
+    end
+
+    # Merges the elements of _e1_ and _e2_ accoring to the implemented strategy
+    #
+    # @param e1 [Object] The first element
+    # @param e2 [Object] The second element
+    # @param merge_options [Hash<String,Object>] Merge options
+    # @return [Object] The result of the merge
+    #
+    def merge(e1, e2, merge_options)
+      checked_merge(
+        assert_type('e1', value_t, e1),
+        assert_type('e2', value_t, e2),
+        assert_type('merge_options', options_t, merge_options))
+    end
+
+    protected
+
+    # Returns the symbolic key identifier for this strategy
+    #
+    # @return [Symbol] The symbolic key
+    #
+    def key
+      raise NotImplementedError, 'Subclass must implement key'
+    end
+
+    # Returns the type used to validate the options hash
+    #
+    # @return [Puppet::Pops::Types::PStructType] the puppet type
+    #
+    def options_t
+      @options_t ||=TypeParser.new.parse("Struct[{strategy=>Optional[Pattern[#{key}]]}]")
+    end
+
+    # Returns the type used to validate the options hash
+    #
+    # @return [Puppet::Pops::Types::PAnyType] the puppet type
+    #
+    def value_t
+      raise NotImplementedError.new('Subclass must implement value_t')
+    end
+
+    def checked_merge(e1, e2, options)
+      raise NotImplementedError.new('Subclass must implement checked_merge(e1,e2,options)')
+    end
+
+    def assert_type(param, type, value)
+      TypeAsserter.assert_instance_of(param, type, value)
+    end
+  end
+
+  # Performs a native Hash merge
```

Comment:
not very much that is "native" about it - better to explain what it does. The fact that ruby hash.merge is used is not teribly interesting - in fact, it does it in the reverse order due to priority of data lookup...